crates/dht/src/time.rs

```diff
@@ -0,0 +1,1034 @@
+//! always of a fixed size, while partial slices are of varying sizes. Full slices occupy
+//! historical time, while partial slices occupy recent time.
+//!
+//! The granularity of time slices is determined by the `factor` parameter. The chosen factor
+//! determines the size of time slices. Where 2^X means "two raised to the power of X", the size of
+//! a full time slice is 2^factor * [UNIT_TIME]. Partial time slices vary in size from
+//! 2^(factor - 1) * [UNIT_TIME] down to 2^0 * [UNIT_TIME] (i.e. [UNIT_TIME]). There is some
+//! amount of time left over that cannot be partitioned into smaller slices.
+//!
+//! Because recent time is expected to change more frequently, combined hashes for partial slices
+//! are stored in memory. Full slices are stored in the Kitsune2 op store.
+//!
+//! The algorithm for partitioning time is as follows:
+//!   - Reserve a minimum amount of recent time as a sum of possible partial slice sizes.
+//!   - Allocate as many full slices as possible.
+//!   - Partition the remaining time into smaller slices. If there is space for two slices of a
+//!     given size, then allocate two slices of that size. Otherwise, allocate one slice of that
+//!     size. Larger slices are allocated before smaller slices.
+//!
+//! As time progresses, the partitioning needs to be updated. This is done by calling the
+//! [PartitionedTime::update] method. The update method will:
+//!    - Run the partitioning algorithm to determine whether new full slices can be allocated and
+//!      how to partition the remaining time into partial slices.
+//!    - Store the combined hash of any new full slices in the Kitsune2 op store.
+//!    - Store the combined hash of any new partial slices in memory.
+//!    - Update the `next_update_at` field to the next time an update is required.
+//!
+//! As an example, consider a factor of 9. This means that full slices are 2^9 = 512 times the [UNIT_TIME].
+//! With a unit time of 15 minutes, that means full slices are 512 * 15 minutes = 7680 minutes = 128 hours.
+//! So every 5 days, a new full slice is created.
+//! The partial slices reserve at least 2^8 + 2^7 ... 2^0 = 519 times the [UNIT_TIME], so 519 * 15 minutes
+//! = 7785 minutes = 129.75 hours. That gives roughly 5 days of recent time.
+//!
+//! A lower factor allocates less recent time and more slices to be stored but is more granular when
+//! comparing time slices with another peer. A higher factor allocates more recent time and fewer
+//! slices to be stored but is less granular when comparing time slices with another peer.
+
+use crate::constant::UNIT_TIME;
+use kitsune2_api::{DynOpStore, K2Error, K2Result, OpId, Timestamp};
+use std::time::Duration;
+
+#[derive(Debug)]
+#[cfg_attr(test, derive(Clone, PartialEq))]
+pub struct PartitionedTime {
+    /// The timestamp at which the time partitioning starts.
+    ///
+    /// This must be a constant among peers who need to compute the same time slices.
+    origin_timestamp: Timestamp,
```

Comment:
My understanding is that the origin time largely only affects the alignment, and that the sizes of the partitions are fully defined by when "now" is. In which case it shouldn't matter how far back in time the origin is.

If that is true, could always use the unix epoch, and eliminate this aspect of complexity?

Reply:
The origin serves as a starting point for the first time slice. Without it, how do you know when the first slice begins? You'd have several empty time slices until the first timestamp of an op. It'd be consistent across peers, but with these empty slices. That's my understanding.

Reply:
I'm not convinced we need to know when the first slice begins. When a gossip round starts we might need to gossip the earliest entry we have for the overlapping arcs, though.

Reply:
I understand that the protocol would be
- host initiates gossip
- remote returns list of time slices with super hashes
- host compares to its own list of time slices and super hashes (first number of time slices, if it differs, add list of ops to fetch queue, if it is equal compare hashes)
- if host or remote does not have list of time slices and super hashes in memory yet, compute them
To compute them, you need to start somewhere, and this somewhere must match on host and remote to arrive at the same slicing. If you start at unix epoch, you'd have a load of time slices at the beginning that contain no ops, because they are before the beginning of the app.

Reply:
The time slices are of fixed size for full slices. So if you have more time, like back to the Unix epoch, then you have to store `[0, 0, ...., 0]` hashes for a bunch of time slices that didn't have a chance of data being in them. My thinking is that when the factor is appropriately chosen, storing a hash for a time range is an acceptable amount of metadata. 

I am very much relying on this per-app constant to reduce the amount of work that needs to be done.

I have a slightly different way I want to go about negotiating changed time slices but that is the right idea yes. You want to both be able to compute the same time slices without prior knowledge of each other's state. Because your earliest op could change, especially early on in a network's life, you'd have to agree that before you could send your summary hashes. That adds an extra round trip to gossip.

Reply:
I feel like it's possible to do this differently, but I don't want to spend any more time on it. Current holochain already uses an origin time, so it will be easier to integrate if it works this way anyways.

Reply:
Yes. There is at least one way. We could just compress runs of 0 hashes.

I was thinking that this might be useful anyway. If your time slices are something like 48 hours and everyone is offline for a week, then you could get some value from compressing those `[0, 0, ..., 0]`.

Then this could just compress everything from the unix timestamp up to whenever we first see a non-empty time slice. That would be simple enough to do I think. There is the possibility that we insert into a compressed section and have to expand it but I can live with that.

It's a touch of extra complexity, but I like that it always applies so that it eliminates the `origin_time` field and solves a general problem

Reply:
This would affect the storage on the host and it would also mean we'd never send lots of 0 hashes over the network, we'd just send one "big" one at the front of the list. Which sounds like it's a constant overhead, but it isn't really if you install an app with a network seed, 6 months after it was packaged. That's just different bits in the "length" of the first block

Reply:
You don't need to send the empty slices' hashes at all if the algorithm is deterministic on both sides, do you? You'd need to store the timestamp of the first actual time slice somehow, if you don't want to calculate it every time anew. So you'd end up with an internal origin time, just not set by app devs. Is that right?

Reply:
I like the combining zeros idea. It's pretty easy to understand, especially since we can give ipv6 string encoding as an example. If you think it's worth considering, would it make the most sense to do that as a separate effort after this PR?

Reply:
Sounds like good support for this, I vote for a follow up change because it'll be easier to review as a change. Yes that means introducing origin_time just to remove it but I think that's okay

Reply:
https://github.com/holochain/kitsune2/issues/32